src/components/editor/alignment.js

```diff
@@ -0,0 +1,58 @@
+'use strict'
+
+const alignCmd = (direction) =>
+  (state, dispatch) => {
+    let tr = state.tr
+    state.doc.nodesBetween(
+      state.selection.from,
+      state.selection.to,
+      (node, pos) => {
+        // align nodes that support alignment
+        if (node.type.attrs.align) {
+          tr.setNodeMarkup(pos, null, { align: direction })
+        }
+      })
+    dispatch(tr)
+  }
+
+const aligns = {
+  left: alignCmd('left'),
+  right: alignCmd('right'),
+  center: alignCmd('center'),
+}
+
+// Maintain alignment and marks by appending the `align` attribute to newly created paragraphs
+// (based on `align` attribute value of the preceding node).
+// Supports both lists and regular markup.
+const keepAlignment = (state, dispatch) => (tr => {
+  const { pos, parentOffset } = state.selection.$cursor
+  const prev = state.doc.resolve(pos - parentOffset)
+  if (prev && prev.parent) {
+    const direction = prev.parent.attrs.align
+    const prevMarks = state.doc.resolve(pos).marks()
+    if (direction && direction !== 'left') {
+      // if we are within a list:
+      if (prev.node(prev.depth - 1).type.name === 'list_item') {
+        // We need to dispatch the previous changes. This will update the UI.
+        dispatch(tr)
```

Comment:
@styts any chance we can do this in a transaction? The thought of having to dispatch multiple times for an alignment change worries me.

Reply:
In the case of the list, the previous commands in `chainCommands` would have created the new list node and paragraph inside it. The state that is passed in does not contain these nodes yet, so it's not possible to `setNodeMarkup` on a node that does not exist. I've tried to do it in one transaction initially, and it works for the regular case, just not in lists.